### DISCUSSION TITLE
Sending parameters which are not defined in the schema

### DISCUSSION THREAD

**Discussion by hegebharat** [schemathesis/schemathesis#1045] [Show and tell]
**Describe the bug**
Thanks a lot for the library. I have scenario where I need to send the additional data as param(query), but which is not defined in the schema. I have to send for every request, Using before_generate_query I sent the request. But since val (query["val"] = "1234222") is not defined in schema I am getting the `TypeError: 'NoneType' object does not support item assignment` error. I think this is because by default query value is set as None.

```python
@schemathesis.hooks.register
def before_generate_query(context, strategy):
    def set_token(query):
        query["val"] = "1234222"
        return query
    return strategy.map(set_token)
```

So Is there any other way can I send the param in the request which is not defined in the schema.
OR based on the endpoint can I run `before_generate_query`?

Please let me know. 

**Environment (please complete the following information):**
 - OS: [e.g. Linux or Windows]
 - Python version: [e.g. 3.7.2]
 - Schemathesis version: [e.g. 2.4.1]
 - Spec version: [e.g. Open API 3.0.2]

Thanks a lot for the help.

**Comment by Stranger6667** [comment]
Hi @hegebharat !

Thanks for asking! :) I think that in this case, the most generic way to do it will be like this:

```python

@schemathesis.hooks.register
def before_generate_query(context, strategy):
    def set_token(query):
        query = query or {}
        query["val"] = "1234222"
        return query
    return strategy.map(set_token)
```

The default value for such cases is `None` to distinguish it from the case where there are optional parameters, but none of them were generated.

Let me know if it helps!

**Reply by Stranger6667** [reply]
Additionally, you can add a condition on the API operation:

```python
@schemathesis.hooks.register
def before_generate_query(context, strategy):
    def set_token(query):
        if context.operation.path == "/whatever-path-you-need/" and context.operation.method == "POST":
            return {"val": "1234222"}
        return query
    return strategy.map(set_token)
```

**Reply by hegebharat** [reply]
Thanks a lot. Its working as expected.
